File: app/meta/process_nav_order.py
"""
ترتیب نمایش فرایندها در سایدبار — اولویت راه‌اندازی اتوماسیون و گردشکار مرکز.

هم‌راستا با موج‌های ۱ و ۲ در docs/institute_onboarding_test_guide_fa.md؛
بقیهٔ فرایندها بر اساس sop_order و در نهایت برچسب فارسی.
"""

from __future__ import annotations

from typing import Any

from app.meta.sop_registry import get_sop_order_for_process_code

# موج ۱ — فرایندهای اصلی (چرخه آموزشی و عملیات روزمره)
_WAVE1_ORDER: tuple[str, ...] = (
    "fall_semester_preparation",
    "winter_semester_preparation",
    "introductory_course_registration",
    "introductory_term_end",
    "comprehensive_course_registration",
    "comprehensive_term_start",
    "comprehensive_term_end",
    "lesson_start_per_term",
    "start_therapy",
    "session_payment",
    "attendance_tracking",
    "supervision_block_transition",
    "class_attendance",
    "thesis_defense_request",
    # خاتمه دروس
    "theory_course_completion",
    "skills_course_completion",
    "group_supervision_course_completion",
    "film_observation_course_completion",
    "film_observation_ta_attendance_completion",
    "live_therapy_observation_course_completion",
    "live_therapy_observation_ta_attendance_completion",
    "live_supervision_course_completion",
    "article_writing_completion",
)

# مسیر ورود مرکز — آماده‌سازی ترم پاییز تا پایان ترم آشنایی
_ONBOARDING_ORDER: tuple[str, ...] = (
    "fall_semester_preparation",
    "introductory_course_registration",
    "introductory_term_end",
)

# موج ۲ — فرایندهای پشتیبانی (کنار مسیر اصلی)
_WAVE2_ORDER: tuple[str, ...] = (
    "violation_registration",
    "educational_leave",
    "full_education_leave",
    "therapy_completion",
    "student_non_registration",
    "extra_session",
    "extra_supervision_session",
    "therapy_session_increase",
    "supervision_session_increase",
    "therapy_session_reduction",
    "supervision_session_reduction",
    "supervision_interruption",
    "unannounced_absence_reaction",
    "unannounced_supervision_absence_reaction",
)

_WAVE1_INDEX = {code: idx for idx, code in enumerate(_WAVE1_ORDER)}
_WAVE2_INDEX = {code: idx for idx, code in enumerate(_WAVE2_ORDER)}
_ONBOARDING_INDEX = {code: idx for idx, code in enumerate(_ONBOARDING_ORDER)}


def onboarding_process_codes() -> tuple[str, ...]:
    """فرایندهای مسیر ورود مرکز (آماده‌سازی → ثبت‌نام آشنایی → پایان ترم)."""
    return _ONBOARDING_ORDER


def process_nav_sort_key(process_code: str, label_fa: str = "") -> tuple:
    """کلید مرتب‌سازی: کمتر = مهم‌تر."""
    code = (process_code or "").strip().lower()
    if code in _WAVE1_INDEX:
        return (0, _WAVE1_INDEX[code], code)
    if code in _WAVE2_INDEX:
        return (1, _WAVE2_INDEX[code], code)
    sop = get_sop_order_for_process_code(code)
    if sop is not None:
        return (2, sop, code)
    label = (label_fa or code).casefold()
    return (3, label, code)


def sort_process_nav_rows(rows: list[dict[str, Any]]) -> list[dict[str, Any]]:
    """مرتب‌سازی ردیف‌های کاتالوگ سایدبار فرایند."""
    return sorted(
        rows,
        key=lambda r: process_nav_sort_key(
            r.get("process_code") or "",
            r.get("label_fa") or r.get("process_name_fa") or "",
        ),
    )

File: admin-ui/src/utils/processNavOrder.js
/**
 * ترتیب نمایش فرایندها در سایدبار — اولویت راه‌اندازی اتوماسیون و گردشکار مرکز.
 * هم‌راستا با app/meta/process_nav_order.py و docs/institute_onboarding_test_guide_fa.md
 */

/** @type {readonly string[]} */
const WAVE1_ORDER = [
  'fall_semester_preparation',
  'winter_semester_preparation',
  'introductory_course_registration',
  'introductory_term_end',
  'comprehensive_course_registration',
  'comprehensive_term_start',
  'comprehensive_term_end',
  'start_therapy',
  'session_payment',
  'attendance_tracking',
  'supervision_block_transition',
  'class_attendance',
  'thesis_defense_request',
]

/** @type {readonly string[]} */
const WAVE2_ORDER = [
  'violation_registration',
  'educational_leave',
  'full_education_leave',
  'therapy_completion',
  'supervision_50h_completion',
  'internship_readiness_consultation',
  'theory_course_completion',
  'student_non_registration',
]

const WAVE1_INDEX = Object.fromEntries(WAVE1_ORDER.map((code, idx) => [code, idx]))
const WAVE2_INDEX = Object.fromEntries(WAVE2_ORDER.map((code, idx) => [code, idx]))

/**
 * @param {string} processCode
 * @param {string} [labelFa]
 * @param {number | null | undefined} [sopOrder]
 */
export function processNavSortKey(processCode, labelFa = '', sopOrder = null) {
  const code = String(processCode || '').trim().toLowerCase()
  if (code in WAVE1_INDEX) return [0, WAVE1_INDEX[code], code]
  if (code in WAVE2_INDEX) return [1, WAVE2_INDEX[code], code]
  if (typeof sopOrder === 'number' && Number.isFinite(sopOrder)) return [2, sopOrder, code]
  return [3, String(labelFa || code), code]
}

/**
 * @param {string} processCode
 * @param {string} [labelFa]
 * @param {number | null | undefined} [sopOrder]
 * @returns {0 | 1 | 2 | 3}
 */
export function processNavUsageTier(processCode, labelFa = '', sopOrder = null) {
  return processNavSortKey(processCode, labelFa, sopOrder)[0]
}

/**
 * @param {Array<{ processCode?: string, process_code?: string, label?: string, label_fa?: string, sop_order?: number | null }>} items
 */
export function sortProcessNavItems(items) {
  const list = Array.isArray(items) ? [...items] : []
  return list.sort((a, b) => {
    const codeA = (a.processCode || a.process_code || '').trim().toLowerCase()
    const codeB = (b.processCode || b.process_code || '').trim().toLowerCase()
    const labelA = (a.label || a.label_fa || '').trim()
    const labelB = (b.label || b.label_fa || '').trim()
    const keyA = processNavSortKey(codeA, labelA, a.sop_order)
    const keyB = processNavSortKey(codeB, labelB, b.sop_order)
    for (let i = 0; i < keyA.length; i += 1) {
      if (keyA[i] < keyB[i]) return -1
      if (keyA[i] > keyB[i]) return 1
    }
    return 0
  })
}
